README.md
# Setup OOB

[![Continuous Integration](https://github.com/vicariousinc/setup-oob/workflows/Continuous%20Integration/badge.svg)](https://github.com/vicariousinc/setup-oob/actions?query=workflow%3AContinuous%20Integration)

This is a utility for configuring out-of-band management systems from within
the running (Linux) OS.

It is built to be highly modular and easy to add support for new types of OOB
devices.

As of current writing it supports SuperMicro's SMC systems and Dell's iDRAC
systems.

It is specifically designed to be easy to run from a Configuration Management
system such as Chef, and thus has a 'check' mode to enable idempotent behavior.

Setup OOB attempts to do as much as possible with the generic `ipmitool`, and
falls back to vendor-specific tools only where necessary. In fact, great
lengths were taken to determine the underlying IPMI commands in order to not
rely on vendor tools.

## General Use

Note that the `--help` message is always the best place to find all the current
options. This section instead attempts to provide and overview of usage, not an
exhaustive list of all options.

You must tell Setup OOB what type of device it needs to configure with the
`--type` option. As of now it takes either `smc` or `drac`. Let's say you want
to configure an iDRAC system, you might check to see if it's configured
properly with:

```shell
setup_oob --type drac --check --name server001-oob --network-mode shared \
  --network-src dhcp
```

This would return 0 if these were all configured this way or non-zero if
something didn't match. If it was not correct, you could converge it to the
desired settings by re-running without `--check`:

```shell
setup_oob --type drac --name server001-oob --network-mode shared \
  --network-src dhcp
```

If configuring a local system, IPMI does not need a password, however for some
devices, when vendor-specific tooling is required, it may fail without
specifying the password via `--old-password` (unless it's still the device
default).

## SMC Licensing

Setup OOB can "activate" SMC systems for you. Automating this can be a bit
difficult without having to store every single license in code or have your
Configuration Management system reach out to an external database.

However, license keys can be derived from the host MAC address and a private
key. *IF YOU HAVE PURCHASED LICENSES* (and if your legal team approves), you
can find the key on the internet and pass it into `--key-file`. If you do this,
then Setup OOB can generate the license key and activate the license for you.

To be clear: *DO NOT DO THIS IF YOU HAVE NOT PURCHASED A LICENSE FOR THE
MACHINES IN QUESTION*. Doing so is certainly a violation of your agreement with
SMC.

The authors of this software are not responsible for illegal use of this
option.

## External tools

Setup OOB requires `ipmitool` at a minimum. For SMC hosts, that is the only
requirement. For DRAC hosts, it also requires racadm.


CHANGELOG.md
# Changelog

## 0.0.2

- Fix gem dependencies

## 0.0.1

- Initial release


lib/setup_oob/oob.rb
# vim: syntax=ruby:expandtab:shiftwidth=2:softtabstop=2:tabstop=2

# Copyright 2021-present Vicarious
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'logger'
require 'socket'
require 'mixlib/shellout'
require_relative 'command/smc'
require_relative 'command/drac'

# A simple class to do the magic to build the right classes and call the
# right methods.
class OOB
  def initialize(config)
    @host = config[:host] || 'localhost'
    @user = config[:user] || user
    @password = config[:password]
    @level = config[:level]
    @desired_hostname = config[:desired_hn] || build_hostname
    logger.debug("Desired hostname set to #{@desired_hostname}")
    @network_mode = config[:network_mode]
    @network_src = config[:network_src]
    @key = config[:key]
    @type = config[:type]
  end

  def logger
    @logger ||= begin
      logger = Logger.new($stdout)
      logger.level = @level
      logger
    end
  end

  def user
    'ADMIN'
  end

  def converged?
    do_work
  end

  def converge!
    do_work(true)
  end

  private

  def do_work(converge = false)
    # Mmmmm, metaprogramming.
    todo = {
      # password MUST come first
      'password' => @password,
      'hostname' => @desired_hostname,
      'ntp' => ['0.pool.ntp.org', '1.pool.ntp.org'],
      'networkmode' => @network_mode,
      'networksrc' => @network_src,
      'ddns' => nil,
    }

    if @type == 'smc'
      if @key
        todo['license'] = @key
      else
        logger.warn('Will not check/activate license, no private key available')
      end
    end

    unless @password
      logger.warn('Will not check/set admin password, no password specified')
      todo.delete('password')
    end

    unless @network_src
      logger.warn('Will not set network_src, not specified')
      todo.delete('networksrc')
    end

    unless @network_mode
      logger.warn('Will not set network_mode, not specified')
      todo.delete('networkmode')
    end

    ret = true
    optionally_supported = ['ntp', 'ddns']
    cmd = Kernel.const_get("#{@type.upcase}Commands")
    todo.each do |name, arg|
      cls = cmd.const_get(name.capitalize)
      obj = cls.new(@host, logger, arg)
      begin
        if converge
          obj.converge!
        else
          ret &= obj.converged?
        end
      rescue Mixlib::ShellOut::ShellCommandFailed => e
        if optionally_supported.include?(name) &&
           e.message.include?('Invalid data field in request')
          logger.warn("Host does not seem to support #{name}, skipping")
          next
        end
        raise
      end
    end

    unless converge
      return ret
    end
  end

  def build_hostname
    if @host == 'localhost'
      "#{Socket.gethostname.split('.').first}-oob"
    else
      @host
    end
  end
end


lib/setup_oob/command/base.rb
# vim: syntax=ruby:expandtab:shiftwidth=2:softtabstop=2:tabstop=2

# Copyright 2021-present Vicarious
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# The base class for a command
class CommandBase
  attr_accessor :logger

  def initialize(host, logger, data)
    @host = host
    @logger = logger
    # data is any extra data subcommands might need
    @data = data
  end

  def converged?
    converged = _converged?
    logger.info("#{pretty_self}: converged: #{converged}")
    converged
  end

  def converge!
    logger.info("Validating #{pretty_self}")
    _converge!
  end

  private

  def _converged?
    fail NotImplementedError
  end

  def _converge!
    fail NotImplementedError
  end

  def pretty_self
    self.class.to_s.split('::').last
  end

  # not applicable to all sub-classes, but common enough might
  # as well include it here.
  def enabled?
    fail NotImplementedError
  end

  # get a single value you from the device
  def getval(_key)
    fail NotImplementedError
  end

  # set key to val
  def setval(_key, _val)
    fail NotImplementedError
  end

  # same as get, but handles multi-value gets and returns a hash
  def getvals(_key)
    fail NotImplementedError
  end

  # Simpler wrapper around Mixlib::ShellOut to log the command
  def run(cmd, forcefail = true)
    logger.debug("Running: #{cmd.join(' ')}")
    s = Mixlib::ShellOut.new(*cmd)
    s.run_command
    s.error! if forcefail
    s
  end

  # Given an array of bytes, get a string out of it.
  def bytes_to_str(bytes)
    bytes.pack('C*').force_encoding('utf-8')
  end

  def basecmd(_defaultpass)
    fail NotImplementedError
  end
end


lib/setup_oob/command/drac.rb
# vim: syntax=ruby:expandtab:shiftwidth=2:softtabstop=2:tabstop=2

# Copyright 2021-present Vicarious
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'mixlib/shellout'
require_relative 'base'
require_relative 'mixins'

# A slight extension of the Command class to add some DRAC-specific
# utility functions. Will be the base class for all DRAC commands
class DRACCommandBase < CommandBase
  def getval(key)
    s = run(basecmd + ['get', key])
    s.stdout.lines[1].strip.split('=')[1]
  end

  def setval(key, val)
    run(basecmd + ['set', key, val])
  end

  def getvals(key)
    s = run(basecmd + ['get', key])
    data = {}
    s.stdout.each_line do |line|
      next if line.start_with?('[')
      next if line.strip.empty?

      k, v = line.strip.split('=')
      data[k] = v
    end
    data
  end

  def basecmd(_defaultpass = false)
    if @host == 'localhost'
      ['racadm']
    else
      fail NotImplementedError
    end
  end
end

# The collection of all DRAC command classes
class DRACCommands
  # Manage the hostname
  class Hostname < DRACCommandBase
    private

    include CommandMixins::Hostname

    def hostname
      getval('iDRAC.NIC.DnsRacName')
      # data = getvals('iDRAC.NIC')
      # "#{data['DNSRacName']}.#{data['DNSDomainName']}"
    end

    def set_hostname
      setval('iDRAC.NIC.DNSRacName', desired_hostname)
      # setval('iDRAC.NIC.DNSDomainName', dn)
    end
  end

  # Manage NTP
  class Ntp < DRACCommandBase
    private

    include CommandMixins::Ntp

    def vals
      @vals ||= getvals('idrac.NTPConfigGroup')
    end

    # Like the super-class, except we have to save the magic bytes...
    def enabled?
      vals['NTPEnable'] == 'Enabled'
    end

    def enable
      setval('idrac.NTPConfigGroup.NTPEnable', 'Enabled')
    end

    def get_server(idx)
      vals["NTP#{idx + 1}"]
    end

    def set_server(idx)
      setval("idrac.NTPConfigGroup.NTP#{idx + 1}", servers[idx])
    end
  end

  class Ddns < DRACCommandBase
    private

    include CommandMixins::Ddns

    def enabled?
      getval('iDRAC.NIC.DNSRegister') == 'Enabled'
    end

    def enable
      setval('iDRAC.NIC.DNSRegister', 'Enabled')
    end
  end

  class Networksrc < DRACCommandBase
    include CommandMixins::Networksrc
  end

  class Networkmode < DRACCommandBase
    private

    include CommandMixins::Networkmode

    def mode_correct?
      logger.debug('  - Checking NIC mode')
      getval('iDRAC.NIC.Selection').downcase == desired_mode
    end

    def set_mode
      case desired_mode
      when 'shared'
        setval('iDRAC.NIC.Selection', 'LOM1')
      when 'dedicated'
        setval('iDRAC.NIC.Selection', desired_mode.capitalize)
      else
        fail "Unknown NIC mode: #{desired_mode}"
      end
    end
  end

  class Password < DRACCommandBase
    private

    include CommandMixins::Password

    def admin_id
      return @admin_id if @admin_id

      s = run(basecmd + ['get', 'iDrac.Users'])
      s.stdout.each_line do |line|
        next unless line.start_with?('iDRAC.Users')

        key = line.split[0]
        info = getvals(key)
        if info['UserName'] == 'root'
          @admin_id = key.split('.')[2]
          return @admin_id
        end
      end
    end
  end
end


lib/setup_oob/command/mixins.rb
# vim: syntax=ruby:expandtab:shiftwidth=2:softtabstop=2:tabstop=2

# Copyright 2021-present Vicarious
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'ipaddress'

# Mixins for the stuff that's common to a given command regardless
# of the platform
module CommandMixins
  module Hostname
    def _converged?
      hn = hostname
      logger.debug("'#{hn}' vs '#{desired_hostname}'")
      hn == desired_hostname
    end

    def _converge!
      unless _converged?
        logger.info(" - Setting hostname (#{desired_hostname})")
        set_hostname
      end
    end

    def desired_hostname
      @data
    end
  end

  module Ntp
    def _converged?
      servers_correct = true
      servers.each_with_index do |server, idx|
        res = get_server(idx) == server
        logger.debug(" - NTP#{idx + 1} correct: #{res}")
        servers_correct &&= res
      end
      enabled? && servers_correct
    end

    def _converge!
      logger.debug(' - Checking if enabled')
      unless enabled?
        logger.info(' - Enabling NTP')
        enable
      end
      servers.each_with_index do |server, idx|
        logger.debug(" - Checking if NTP#{idx + 1} is correct")
        unless get_server(idx) == server
          logger.info(" - Setting NTP#{idx + 1} server")
          set_server(idx)
        end
      end
    end

    def servers
      @data
    end
  end

  module Networkmode
    def _converged?
      mode_correct?
    end

    def desired_mode
      @data
    end

    def _converge!
      unless mode_correct?
        logger.info(" - Setting network to #{desired_mode}")
        set_mode
      end
    end
  end

  module Ddns
    def _converged?
      enabled?
    end

    def _converge!
      unless enabled?
        logger.info(' - Enabling DDNS')
        enable
      end
    end
  end

  module Networksrc
    def desired_mode
      @data == 'dhcp' ? 'dhcp' : 'static'
    end

    def desired_address
      if @data == 'dhcp'
        fail 'Set to DHCP, but looking for address, what?'
      end

      IPAddress(@data).address
    end

    def desired_netmask
      if @data == 'dhcp'
        fail 'Set to DHCP, but looking for address, what?'
      end

      IPAddress(@data).netmask
    end

    def mode
      x = current['IP Address Source']
      case x
      when /DHCP/
        'dhcp'
      when /Static/
        'static'
      else
        'other'
      end
    end

    def address
      current['IP Address']
    end

    def netmask
      current['Subnet Mask']
    end

    def current
      return @current if @current

      @current = {}
      cmd = ipmicmd + ['lan', 'print', '1']
      s = run(cmd)
      s.stdout.each_line do |line|
        key, val = line.chomp.split(':')
        @current[key.strip] = val.strip
      end
      @current
    end

    def _converged?
      mode? && address?
    end

    def address?
      if desired_mode == 'static'
        logger.debug('  - Checking of address is set')
        logger.debug("'#{@data}' vs '#{@data}'")
        return address == desired_address || netmask == desired_netmask
      end
      true
    end

    def mode?
      logger.debug("  - Checking if network src mode set to #{desired_mode}")
      mode == desired_mode
    end

    def _converge!
      unless mode?
        logger.info(" - Setting network src to #{desired_mode}")
        set_mode
      end
      if desired_mode == 'static'
        logger.info(' - Checking address')
        if address != desired_address
          logger.info(" - Setting network address to #{desired_address}")
          set_address
        end
        if netmask != desired_netmask
          logger.info(" - Setting network mask to #{desired_netmask}")
          set_netmask
        end
      end
    end

    def smc?
      @smc ||= self.class.to_s.start_with?('SMC')
    end

    def set_mode
      cmd = ipmicmd(true) + ['lan', 'set', '1', 'ipsrc', desired_mode]
      run(cmd)
    end

    def set_address
      cmd = ipmicmd(true) + ['lan', 'set', '1', 'ipaddr', desired_address]
      run(cmd)
    end

    def set_netmask
      cmd = ipmicmd(true) + ['lan', 'set', '1', 'netmask', desired_netmask]
      run(cmd)
    end

    def ipmicmd(set = false)
      smc? ? basecmd(set) : SMCCommandBase.basecmd('localhost')
    end
  end

  module Password
    def password
      @data
    end

    def _converged?
      password_set?
    end

    def _converge!
      unless password_set?
        logger.info(' - Setting password')
        set_password
      end
    end

    def smc?
      @smc ||= self.class.to_s.start_with?('SMC')
    end

    # It turns out that 'ipmitool user test' and 'ipmitool user set'
    # work fairly unversally. And there's no way to do 'user test'
    # in racadm. So this should work pretty much everywhere.
    #
    # Though... 'ipmitool user list', for some reason, is not so
    # universal. Womp Womp.
    def password_set?
      id = admin_id
      cmd = ipmicmd + ['user', 'test', id, smc? ? '20' : '16', password]
      s = run(cmd, false)
      !s.error?
    end

    def set_password
      id = admin_id
      cmd = ipmicmd(true) + ['user', 'set', 'password', id, password]
      s = run(cmd)
      if s.error?
        # if the password isn't what we expect *and* isn't ADMIN, it may be
        # serial or an old password. We can reset it to ADMIN
        # NOTE: this actually resets ALL it's settings. Which is OK because
        # we set password first, so the rest will get set properly
        run(ipmicmd(true) + ['raw', '0x30', '0x48', '0x1'])
        # takes about  seconds for it to come to its senses
        sleep(5)
        # now, try again
        run(cmd)
      end
    end

    def ipmicmd(set = false)
      smc? ? basecmd(set) : SMCCommandBase.basecmd('localhost')
    end
  end
end


lib/setup_oob/command/smc.rb
# vim: syntax=ruby:expandtab:shiftwidth=2:softtabstop=2:tabstop=2

# Copyright 2021-present Vicarious
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

require 'openssl'
require_relative 'base'
require_relative 'mixins'

# A slight extension of the Command class to add some SMC-specific
# utility functions. Will be the base class for all SMC commands
class SMCCommandBase < CommandBase
  DEFAULT_PASSWORD = 'ADMIN'.freeze
  USER = 'ADMIN'.freeze
  COMMANDS = {
    :hostname => [0x47],
    :ntp => [0x68, 0x01],
    :ddns => [0x68, 0x04],
    :networkmode => [0x70, 0x0c],
    :isactivated => [0x6A],
    :setlicense => [0x69],
    # some magic set of bytes for firmware version and mac ...
  }.freeze

  ACTIONS = {
    :get => [0x00],
    :set => [0x01],
  }.freeze

  def enabled?
    # assumes a sub-classed cmdbytes
    data = cmdbytes(:get, :enabled)
    out = runraw(data)
    # if the first byte is 1, "it" is enabled, whatever "it" is
    en = out[0] == 1
    logger.debug("#{pretty_self} enabled: #{en}")
    en
  end

  # Get the IPMI bytes for the command we want to run
  def cmdbytes(command, action = nil)
    data = [0x30] + COMMANDS[command]
    if action
      data += ACTIONS[action]
    end
    data
  end

  # Wrapper to build the "raw" command, run it, and convert the answer into
  # an array of actual integers.
  def runraw(data)
    s = run(rawcmd(data))
    s.stdout.split.map { |x| x.to_i(16) }
  end

  # Builds the 'raw' command using 'data'
  def rawcmd(data)
    basecmd + ['raw'] + data.map(&:to_s)
  end

  def basecmd(defaultpass = false)
    SMCCommandBase.basecmd(@host, @user, defaultpass ? 'ADMIN' : @password)
  end

  def self.basecmd(host, user = nil, pass = nil)
    if host == 'localhost'
      ['ipmitool']
    else
      unless user && pass
        fail 'basecmd: No user and password sent with host'
      end

      [
        'ipmitool',
        '-H', host,
        '-U', user,
        '-P', pass
      ]
    end
  end
end

# The collection of all SMC command classes
class SMCCommands
  class Hostname < SMCCommandBase
    private

    include CommandMixins::Hostname

    def desired_hostname
      @data
    end

    def hostname
      # for hostname get is 0x02 and set is 0x01. no idea why
      data = cmdbytes(:hostname) + [0x02]
      out = runraw(data)
      bytes_to_str(out)
    end

    def set_hostname
      # no terminating null for this command...
      data = cmdbytes(:hostname) + [0x01] + desired_hostname.bytes
      runraw(data)
    end
  end

  class Ntp < SMCCommandBase
    SUB_COMMANDS = {
      :enabled => [0x00],
      :primary => [0x01],
      :secondary => [0x02],
    }.freeze

    TYPES = [
      :primary,
      :secondary,
    ].freeze

    private

    include CommandMixins::Ntp

    def cmdbytes(action, subcmd)
      super(:ntp, action) + SUB_COMMANDS[subcmd]
    end

    def enabled?
      data = cmdbytes(:get, :enabled)
      out = runraw(data)
      en = out[0] == 1
      # when you check if NTP is enabled, a bunch of extra bytes
      # are passed back that MUST be passed in when enabling NTP
      @_magic = out[1..-1]
      logger.debug("NTP enabled: #{en}, magic bytes: #{@_magic}")
      en
    end

    def enable
      # 0x01 is "enable"
      data = cmdbytes(:set, :enabled) + [0x01] + @_magic
      runraw(data)
    end

    def get_server(idx)
      data = cmdbytes(:get, type(idx))
      bytes_to_str(runraw(data))
    end

    def type(idx)
      TYPES[idx]
    end

    def set_server(idx)
      name_bytes = servers[idx].bytes
      data = cmdbytes(:set, type(idx)) + name_bytes + [0x00] # null termination
      bytes_to_str(runraw(data))
    end
  end

  class Networksrc < SMCCommandBase
    include CommandMixins::Networksrc
  end

  class Networkmode < SMCCommandBase
    private

    MODES = {
      :dedicated => 0x00,
      :shared => 0x01,
      :failover => 0x02,
    }.freeze

    include CommandMixins::Networkmode

    def mode_val
      val = MODES[desired_mode.to_sym]
      unless val
        fail "No such mode #{desired_mode}"
      end

      val
    end

    def cmdbytes(action, _subcmd = nil)
      super(:networkmode, action)
    end

    def mode_correct?
      data = cmdbytes(:get)
      out = runraw(data)[0]
      logger.debug("mode: #{out}")
      out == mode_val
    end

    def set_mode
      data = cmdbytes(:set) + [mode_val]
      runraw(data)
    end
  end

  class Ddns < SMCCommandBase
    SUB_COMMANDS = {
      :enabled => [0x00],
    }.freeze

    private

    include CommandMixins::Ddns

    def cmdbytes(action = nil, type = nil)
      # this one is backwards...  I dunno why
      data = super(:ddns)
      if action && type
        data += SUB_COMMANDS[type] + ACTIONS[action]
      end
      data
    end

    def enable
      # this seems to put it in some sort of setting mode...
      data = cmdbytes(:set, :enabled)
      runraw(data)
      data = cmdbytes
      # then you do this crazy magic...
      data += [0x01, 0x01, 0x00, 0x7F, 0x00, 0x00, 0x01] +
        '#host.#domain'.bytes + [0x00]
      runraw(data)
    end
  end

  # Manage the licenses
  #
  # READ CAREFULLY!!
  # Generating licenses if you haven't purchased them is illegal!
  # This is here for convenience since the license is derivable from
  # the MAC address, if you have the keys. This allows activating
  # the license in an automated fashion.
  #
  # I make no claims about the legality of having the key, it probably
  # depends on what country you are in. The key is not distributed with
  # this software. But you can find it if you want.
  #
  class License < SMCCommandBase
    def key
      @data
    end

    def _converged?
      licensed?
    end

    def converge!
      unless licensed?
        logger.info(' - Setting license')
        set_license
      end
    end

    private

    # Thanks Peter Kleissner for documenting this.
    def generate_license
      digest = OpenSSL::Digest.new('sha1')
      data = mac.split(':').map { |x| x.to_i(16).chr }.join
      raw = OpenSSL::HMAC.digest(digest, key, data)
      raw_lic = raw.chars[0..11].map(&:ord)
      fmt = "%02X%02X-%02X%02X-%02X%02X-%02X%02X-%02X%02X-%02X%02X\n"
      logger.debug("Generated license: #{raw_lic}")
      logger.debug("Human readable license: #{fmt % raw_lic}")
      raw_lic
    end

    def licensed?
      data = cmdbytes(:isactivated)
      s = runraw(data)
      # There's only one byte, and it's non-zero if we're activated
      en = s[0].positive?
      logger.debug("Activated: #{en}")
      en
    end

    def mac
      # does not use `cmdbytes` because this is the raw sequence of
      # bytes and not inside of the "0x30' netfn that other commands
      # are
      data = [0x0C, 0x02, 0x01, 0x05, 0x00, 0x00]
      # we want actual hex strings, so use run directly
      s = run(rawcmd(data))
      hex = s.stdout.split
      # first byte is version, I think
      hex.shift
      mac = hex.join(':')
      logger.debug("Mac is #{mac}")
      mac
    end

    def set_license
      lic = generate_license
      data = cmdbytes(:setlicense) + lic
      s = runraw(data)
      unless s[0].zero?
        fail 'Failed to set license key'
      end
    end
  end

  class Password < SMCCommandBase
    private

    include CommandMixins::Password

    # Find the user id of the ADMIN user. Usually 2, but I
    # didn't want to hard-code that.
    def admin_id
      return @admin_id if @admin_id

      s = run(basecmd + ['user', 'list'])
      s.stdout.each_line do |line|
        next if line.start_with?('ID')

        bits = line.split
        if bits[1] == 'ADMIN'
          @admin_id = bits[0]
          return @admin_id
        end
      end
    end
  end
end
